File: Forest/cgNode.cpp
#include "cgNode.h"



cgNode* cgNode::clone_node(cgNode* node)
{
	if (!node)	return nullptr;
	if (node == nullNode::get())	return nullNode::get();
	Functor* fc = nullptr;
	Diffor* dc = nullptr;
	Trimor* tc = nullptr;
	if (node->func)	fc = node->func->clone();
	if (node->diffor) {
		dc = node->diffor->clone(nullptr);
	}
	if (node->trimor)	tc = node->trimor->clone(nullptr);
	cgNode* res = new cgNode(clone_node(node->left), clone_node(node->right), fc, dc, tc, node->val);
	res->diffor->p = res;
	res->trimor->p = res;
	return res;
}

cgNode::cgNode(cgNode* l, cgNode* r, Functor* f, Diffor* d, Trimor* t, const double& v = 0.0)
	: left(l), right(r), func(f), diffor(d), trimor(t), val(v) {}

cgNode::~cgNode() {
	if (func)	delete func;
	if (diffor)	delete diffor;
}

cgNode* cgNode::clone()
{
	return clone_node(this);
}

bool cgNode::is_leaf(cgNode* node)
{
	return !(node->left || node->right);
}

void cgNode::set_val(const double& d)
{
	val = d;
}

bool cgNode::is_const()
{
	return false;
}

File: JavaForest/src/cgNode/cgNode.java
package cgNode;

import org.jetbrains.annotations.NotNull;

public class cgNode {
    protected double _val;
    protected cgNode _left;
    protected cgNode _right;
    protected cgNode _parent;

    public cgNode(){
        _right = null;
        _left = null;
        _val = 0;
    }

    public cgNode(cgNode parent, cgNode left, cgNode right, double val){
        _val = val;
        _left = left;
        _right = right;
        _parent = parent;
    }

    public cgNode(@NotNull cgNode node){
        _val = node._val;
        _left = node._left;
        _right = node._right;
        _parent = node._parent;
    }

    public double compute(double n1, double n2){
        return _val;
    }

    public cgNode trim(){
        return this;
    }

    public cgNode diffor(){
        return this;
    }

    public cgNode clone() throws CloneNotSupportedException {
        return new cgNode(this);
    }
}
